### ISSUE TITLE
Understanding megalodon output in IGV

### ISSUE THREAD

**Issue by danrdanny** [nanoporetech/megalodon#212]
Hi, and thanks for the effort developing megalodon. We're working on incorporating it into our analysis pipeline and have gotten stuck on understanding what the right way to view the output in IGV is. Would appreciate advice on what the output should look like.

Reads generated on PromethION, R9.4.1 flowcell
megalodon version 2.3.4
IGV version 2.11.3
Base modification probability threshold in IGV: 0.8

Command to generate data:
```
megalodon \
    /path/to/fast5_dir/ \
    --outputs basecalls mappings mod_mappings mods \
    --reference hg38.no_alt.fa \
    --mod-map-emulate-bisulfite \
    --devices 1 --processes 30 \
    --guppy-server-path /path/to/ont-guppy_5.0.7/bin/guppy_basecall_server \
    --output-directory sampleOut-try2 \
    --overwrite
```

A modified bam called `mod_mappings.5mC.bam` is made and can be viewed in IGV. When I view it without changing colors I see this:

<img width="473" alt="image" src="https://user-images.githubusercontent.com/3186059/140699109-38dec92b-9559-4f3e-bee1-cbbc62f3fb69.png">

When I then color alignments by 'bisulfite mode' -> 'CG' I see this:

<img width="438" alt="image" src="https://user-images.githubusercontent.com/3186059/140697853-30dfd994-467f-40fb-9b80-b863a4f40c31.png">

It looks like bases that are predicted to be methylated have been changed to 'M' while the unmethylated ones are red.

When I color by "base modification" I see this:

<img width="450" alt="image" src="https://user-images.githubusercontent.com/3186059/140698747-dd82f47a-1eb9-45fb-bdaa-53e9f48b0077.png">

Are these views a consequence of how I ran megalodon or are they due to an IGV setting? Am I misunderstanding the use of the `mod_mappings.5mC.bam` file? Thanks again.

**Comment by marcus1487** [comment]
TL;DR don't use bisulfite mode for nanopore methylation reads.

I think this is more of a question to post on IGV, but will try to shed some light on the issues. So the main one is that bisulfite is a very specific data type not applicable to nanopore data (unless one were to sequence bisulfite treated data with nanopore and align with a bisulfite-compatible aligner). Bisulfite converts any un-methylated C to G and leaves methylated Cs as Cs. I would separate your visualizations by strand to help see the trends, but the idea is that any canonical C in a read matching a reference C in bisulfite-treated data indicates methylation. It looks like the nanopore methylation tags are taking precedent over the bisulfite C bases (thus M bases are not colored red). Again sorting by strand should help here.

I'm not sure why the color by "base modifications" is not adding colored bars in the coverage section. This may be a design choice by the IGV developers. @jrobinso any thoughts?

**Comment by jrobinso** [comment]
@marcus1487 You are correct about "bisulfite mode", that is for bisulfite-treated experiments as you note.

WRT coloring coverage bar, I remind you of this thread https://github.com/igvteam/igv/issues/991.   If you think the data above should show colors in the coverage bar but isn't, I will need a test dataset to look further.   Its impossible for me to jump into and out of this topic every 6 months and keep any context so I might have missed the point here.

**Comment by danrdanny** [comment]
Thanks @jrobinso and @marcus1487. Totally understand about not using bisulfite mode--just showing the example since I think many people would try that out of habit and experience with other datasets.

@jrobinso, thanks for pointing back to https://github.com/igvteam/igv/issues/991. I did see that and did set the base modification probability threshold to 0.8. I think it's important to point out that people who find this thread should upgrade IGV and check that setting if having problems.

I uploaded a small (2Mb) test dataset (aligned to GRCh38).
bam: https://drive.google.com/file/d/1QvQpN2xbJGX7q5PqobOmxQRV5wiFOVsJ/view?usp=sharing
index: https://drive.google.com/file/d/12OSWWIuflomC6xLeuHm47DBStGd5IPkh/view?usp=sharing

Big picture: people using megalodon might expect a file called `mod_mappings.bam` to look similar to what you see when using bisulfite mode (for example, you may call methylation with nanopolish and then convert a bam file using nanopore-methylation-utilities). My sense is that this is not the case, and this may surprise people. I also could be wrong about what to expect in the `mod_mappings.bam` file. My goal here is to clarify what we should expect to see. I realize you are very busy with development and I appreciate the time.

The file I uploaded contains adaptive sampling reads around XYLT1. There is a CpG island in the 5' UTR that should be unmethylated. When I load `mod_mappings.bam` into IGV and color by 'base modifications' I see this (IGV view is approximately chr16:17,467,532-17,473,840):

![image](https://user-images.githubusercontent.com/3186059/140979311-5a0f3f70-809a-4e7b-a7b7-edcfe992e474.png)

Note that the track with the green block at the bottom is the CpG island. This looks like no methylation over the CpG, as expected. 

When I zoom in, I see very few bases were converted to "M", consistent with there being no methylation detected. This is different than I initially expected--the methylated region is "M".

![image](https://user-images.githubusercontent.com/3186059/140979641-81404dc7-c3ef-4229-8a62-a649f106689d.png)

If I go to a different region, not within the CpG island I see a group of CpGs that appear methylated. IGV view is chr16:17,472,329-17,472,387:

![image](https://user-images.githubusercontent.com/3186059/140980241-e0cca1c4-fb45-4b25-81c1-37b794725792.png)

I just want to make sure this is what is expected. Unmethylated bases aren't converted to an "M", while methylated bases are converted.

I think the challenging thing about this view is that unmethylated bases don't stand out like they do when viewing by bisulfite mode. In a way, it's actually more useful to view in bisulfite mode because then bases that aren't converted are changed to red and can be visualized. It's just not easy to see the difference between the red color and the "M":

![image](https://user-images.githubusercontent.com/3186059/140981709-5218a5d8-419d-4e7b-b92a-c779cdc36ec6.png)

It's fine if this is the expected behavior, I just want to clarify that it is.

Any suggestions on better ways to view the data from anyone are appreciated. I'm happy to open an issue with IGV.

Thanks!

**Comment by jrobinso** [comment]
Yes I see the issue, but if this is the goal... 

_"Big picture: people using megalodon might expect a file called mod_mappings.bam to look similar to what you see when using bisulfite mode (for example, you may call methylation with nanopolish and then convert a bam file using nanopore-methylation-utilities)."_  

then that was not what was implemented.  This feature was driven by input from ONT developers primarily, I'm neutral.

Looking at the code, modified bases are colored gray, which is what you are seeing,  I don't recall the reason for that but there are issue threads on IGV where this was all discussed.   It could be any color we want, however it won't address what you are getting at.   The modified bases mode just colors the modified bases, it does not interpret the specific experiment the way bisulfite mode does.   It is generic for any modification, presumably there are other experiment types that modify bases.    If we want a bisulfite equivalent mode for specific modification types that is another thing and will require development.   That code (bisulifte) is complex, maybe unnecessarily so, but it handles all sorts of contexts not just CG.

**Comment by danrdanny** [comment]
Makes sense @jrobinso, thanks. I'll close this issue and hopefully others will find the thread useful.

**Comment by jrobinso** [comment]
@danrdanny I was just trying to provide some background on the current view, not defend it as the best possible.   If the majority of the community using this are studying DNA methylation we are possibly doing the wrong thing here by keeping it generic,  but that's how we got to this point.   Please feel free to open an issue on the IGV forum to discuss further improvements,  I am not wedded to the current scheme.

**Comment by danrdanny** [comment]
@jrobinso, I may have answered my own question by finding an option I wasn't using in megaldon:

```
--mod-map-emulate-bisulfite \
--mod-map-base-conv C T --mod-map-base-conv m C
```

So, this may not be an issue at all. I'll update this thread once it runs to see if it gives me the output I expected.

**Comment by marcus1487** [comment]
I would recommend that the bisulfite emulation mode is not the long term solution for this issue. We are working hard to improve our 5mC+5hmC model and even more modified bases in the future. And the bisulfite emulation is not compatible with multiple modified base output. If there are visualization features for modified bases I would agree with @jrobinso that if the bisulfite visualization is preferred then updates to the modbam visualization should be requested in igv.

**Comment by danrdanny** [comment]
Agree that bisulfite emulation is probably not a long-term solution. I can make a request in IGV. But, to close this issue out, I'll say that the following commands gave me the view using the `mod_mappings.5mC.bam` file that I initially wanted:

```
--mod-map-emulate-bisulfite \
--mod-map-base-conv C T --mod-map-base-conv m C
```
